### PR TITLE
Fix fast reader tests and warnings for "win32"

### PR DESCRIPTION
**Pull request by dhomeier**
### Description
Follow-up on some issues found in https://github.com/astropy/astropy/pull/15235#discussion_r1308774180

1. The "win32" job uses a 64 bit compiler and as such passes the double precision test, just shows somewhat deviant behaviour on warnings.
2. A construct such as https://github.com/astropy/astropy/blob/53188c39a23c33b72df5850ec59e31886f84e29d/astropy/io/ascii/tests/test_c_reader.py#L1401-L1410
would have never worked for 0 warnings, as `pytest.warns()` already raises an error in that case – using a conditional context here now as in `test_data_out_of_range`.

Note: could just not get this to work using `with pytest.warns(Warning) as result` as suggested in the [pytest docs](https://docs.pytest.org/en/latest/how-to/capture-warnings.html#custom-failure-messages)) as an error is still raised already at that point. Open to suggestions if this could be made to work somehow, but my reading is that the pytest docs are simply wring here.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by pllim — COMMENTED** [review]
I am going to also run devinfra job on this one to make sure the changes here are compatible with pytest dev that is even stricter on warning handling.

I don't grok the purpose of those tests and why they are different on Windows or if that should even happen, so the rest of `io.ascii` maintainers (@taldcroft and @hamogu) should review this carefully.

Thanks!

File: astropy/io/ascii/tests/test_c_reader.py
Comment (pllim) [inline astropy/io/ascii/tests/test_c_reader.py:1418]: No need to shorten `warning_lines` to `w`?

**Comment by hamogu** [comment]
I don't understand either, but it's looks good to me.

File: astropy/io/ascii/tests/test_c_reader.py
Comment (dhomeier) [reply astropy/io/ascii/tests/test_c_reader.py:1418]: Got pestered by black on this earlier; anyway the other tests are also just using `w` (and we get at most a single line).

**Review by mhvk — APPROVED** [review]
I'm slightly wondering why astropy tests should be covering oddities of compilers, but given that we do, this seems the right solution, so approved!

**Comment by pllim** [comment]
Thanks!

**Comment by pllim** [comment]
If anyone ever able to track down exactly what is going on over at Windows for this test, please report back!

**Pull request merged**
